### PR TITLE
updates execute to accept name of contract not address

### LINKED ISSUE #1012: Support sozo execute by name
Currently, it is required to pass the contract address when calling `sozo execute`. We would like to support executing by contract name.

Currently:

```
sozo execute --manifest-path examples/ecs/Scarb.toml 0x19cbafa838c97f28b2392c1b77acf131d8b915f3c892379d7dff501aa4695fc move --calldata 0x314d21c89db29f82669c54f62e291b8b90effc5725fb12291566635b54b8614 --calldata 0x1
```

Where `0x19cbafa838c97f28b2392c1b77acf131d8b915f3c892379d7dff501aa4695fc` is the contract name.

Future:

```
sozo execute --manifest-path examples/ecs/Scarb.toml player_actions move --calldata 0x314d21c89db29f82669c54f62e291b8b90effc5725fb12291566635b54b8614 --calldata 0x1
```

To do so, we need to compute the contract address based on the contract name, which can be done by:

Fetch the base contract class hash from here: https://github.com/dojoengine/dojo/blob/e2082ec8dea2fa58e0dcbca72ca438b8fc6412fe/crates/dojo-core/src/world.cairo#L35
The contract address is calculated like this: https://github.com/dojoengine/dojo/blob/fc2d3aa3e671cd7b7de3c63e585d2865a2340569/crates/dojo-world/src/migration/mod.rs#L177
Where the salt is calculated like this https://github.com/dojoengine/dojo/blob/fc2d3aa3e671cd7b7de3c63e585d2865a2340569/crates/dojo-world/src/migration/strategy.rs#L191
So we only need one call to the world contract to get the base class hash then can compute any contracts address based on its name

This address computation already exists in the code base but can potentially be better abstracted for reuse.

### PR DESCRIPTION
**Pull request by ponderingdemocritus**
- Updates `sozo execute` to accept the name of the contract rather than the address

closes #1012

### CODE REVIEW

File: crates/sozo/src/commands/execute.rs [hunk 1]
```diff
@@ -12,7 +12,7 @@ use crate::ops::execute;
 #[command(about = "Execute a system with the given calldata.")]
 pub struct ExecuteArgs {
     #[arg(help = "The address of the contract to be executed.")]
-    pub contract: FieldElement,
+    pub contract_name: String,
```
Comment (tarrencev) [inline crates/sozo/src/commands/execute.rs:15]: we should support either here, so if it is a `0x` treat it as an address, else as a name

File: crates/sozo/src/ops/execute.rs [hunk 2]
```diff
@@ -1,21 +1,69 @@
...
+        Some(env) => match env.world_address {
+            Some(ref address) => Some(address.clone()),
+            None => None,
+        },
+        None => return Err(anyhow::anyhow!("No World Address found")),
+    };
+
+    let contract_address_str = world_address.as_ref().unwrap().as_str();
+
+    print!("{}", contract_address_str);
```
Comment (tarrencev) [inline crates/sozo/src/ops/execute.rs]:
```suggestion
```

File: crates/sozo/src/ops/execute.rs [hunk 3]
```diff
@@ -1,21 +1,69 @@
...
 
     let provider = starknet.provider(env_metadata.as_ref())?;
 
+    let world_address = match env_metadata.as_ref() {
+        Some(env) => match env.world_address {
+            Some(ref address) => Some(address.clone()),
+            None => None,
+        },
+        None => return Err(anyhow::anyhow!("No World Address found")),
+    };
```
Comment (tarrencev) [inline crates/sozo/src/ops/execute.rs:25]: We can simplify to something like:

```suggestion
let world_address = env_metadata.as_ref()
    .and_then(|env| env.world_address.as_ref())
    .map(|address| address.clone())
    .ok_or_else(|| anyhow::anyhow!("No World Address found"))?;
```

File: crates/sozo/src/ops/execute.rs [hunk 4]
```diff
@@ -1,21 +1,69 @@
...
+    let contract_address_str = world_address.as_ref().unwrap().as_str();
+
+    print!("{}", contract_address_str);
+
+    let contract_class_hash = provider
+        .call(
+            FunctionCall {
+                contract_address: FieldElement::from_hex_be(contract_address_str).unwrap(),
+                entry_point_selector: selector!("base"),
+                calldata: Vec::new(),
```
Comment (tarrencev) [inline crates/sozo/src/ops/execute.rs:34]: I think you can do this like you're doing below:

```suggestion
                calldata: &[],
```

File: crates/sozo/src/ops/execute.rs [hunk 5]
```diff
@@ -1,21 +1,69 @@
...
+    };
+
+    let contract_address_str = world_address.as_ref().unwrap().as_str();
+
+    print!("{}", contract_address_str);
+
+    let contract_class_hash = provider
+        .call(
+            FunctionCall {
+                contract_address: FieldElement::from_hex_be(contract_address_str).unwrap(),
```
Comment (tarrencev) [inline crates/sozo/src/ops/execute.rs:32]: Hmm i think you `world_address` is a felt already?

File: crates/sozo/src/ops/execute.rs [hunk 6]
```diff
@@ -1,21 +1,69 @@
...
+                entry_point_selector: selector!("base"),
+                calldata: Vec::new(),
+            },
+            &BlockId::Tag(BlockTag::Latest),
+        )
+        .await?;
+
+    let salt = poseidon_hash_many(
```
Comment (tarrencev) [inline crates/sozo/src/ops/execute.rs:40]: we have a `generate_salt` helper here https://github.com/dojoengine/dojo/blob/4ffeff5a335f742957d8d167473f156b290629bd/crates/dojo-world/src/migration/strategy.rs#L231

File: crates/sozo/src/ops/execute.rs (hunk 5 above)
Comment (ponderingdemocritus) [reply crates/sozo/src/ops/execute.rs:32]: it's coming from the `Scarb` so it's a string

**Review by tarrencev — APPROVED** [review]

File: crates/sozo/src/ops/execute.rs [hunk 7]
```diff
@@ -10,12 +14,40 @@ pub async fn execute(args: ExecuteArgs, env_metadata: Option<Environment>) -> Re
...
+    let contract_class_hash = provider
+        .call(
+            FunctionCall {
+                contract_address: FieldElement::from_hex_be(&world_address).unwrap(),
+                entry_point_selector: selector!("base"),
+                calldata: [].to_vec(),
+            },
+            BlockId::Tag(BlockTag::Latest),
+        )
+        .await?;
```
[the remark this answers is not in the document]
Comment (ponderingdemocritus) [reply crates/sozo/src/ops/execute.rs:32]: yes sensei

**Pull request merged**